core/trino-main/src/main/java/io/trino/operator/output/PositionsAppenderPageBuilder.java

```diff
@@ -144,6 +144,21 @@ public boolean isEmpty()
         return declaredPositions == 0;
     }
 
+    public boolean shouldFlushBeforeRelease()
+    {
+        if (declaredPositions == 0) {
+            return false;
+        }
+        for (UnnestingPositionsAppender positionsAppender : channelAppenders) {
+            if (positionsAppender.shouldForceFlushBeforeRelease()) {
+                // dictionary encoding will be preserved, so force the current page to be flushed
+                return true;
+            }
+        }
+        // flush the current page if the builder is now full as a result of transitioning dictionaries to direct mode
```

Comment:
> as a result of transitioning dictionaries to direct mode

I think we need to simplify this. It's really odd that `shouldForceFlushBeforeRelease()` has a side effect of flattening block (better name would be `flushOrFlattenIfNeccecery`).

But I think it would be better to call `flatten` here explicitly.

Reply:
Separated `shouldForceFlushBeforeRelease()` from `flattenPendingDictionary()` and called those as appropriate here.